### DISCUSSION TITLE
Why passing ref={} ref={} twice works on a component?

### DISCUSSION THREAD

**Discussion by Grabber** [solidjs/solid#1778] [Q&A]
@ryansolid

I'm trying to implement a chat-like app where the scroll starts at the end. The other feature is a button (scroller) to allow jumping from any position to the bottom of the chat history.

I need to pass the ref from the parent to the child and use it from both. I created a reference both inside the parent and the child (viewportChatMessagesRef). BUT I'm getting the correct HTML object from the parent and just a ref() function inside the child (instead the HTML object).

I tried to create a local ref inside the child and passed two references at once: ref={props.ref} and ref={viewportChatMessagesRef}... somehow it's working, BUT it's very strange! Does "ref={}" concatenate references?

Could you please help me figure what what I'm missing or if we have a bug here?

I read somewhere that forwardRef is not availaible anymore?

```
export default function Chat(props) {
  const [visible, setVisible] = createSignal(true);
  const [scroller, setScroller] = createSignal(false);

  let viewportChatMessagesRef;

  function scrollToBottom() {
    viewportChatMessagesRef.scrollIntoView();
  }

  function handleOnScroll(event) {
    const scroll = event.target.scrollTop / (
                   event.target.scrollHeight - event.target.clientHeight);
    setScroller(scroll < 0.8);
  }

  return (
    <>
    <Show when={visible()}>
      <div class={styles['chat']} onScroll={handleOnScroll}>
        <ViewportChatMessages ref={viewportChatMessagesRef} />

        { scroller() && <ViewportChatScroller scroll={scrollToBottom} /> }
      </div>
    </Show>
    </>
  );
}
```

```
export default function ViewportChatMessages(props) {
  const [visible, setVisible] = createSignal(true);
  const [messages, setMessages] = createSignal([]);

  const DynamicChatMesssagesComponents = {
    l: ViewportChatMessagesLeft,
    c: ViewportChatMessagesCenter,
    r: ViewportChatMessagesRight
  }

  let viewportChatMessagesRef;

  onMount(() => {
    setMessages(messages__1_2_4_8);
  });

  createEffect(on(messages, (v) => {
    setTimeout(() => viewportChatMessagesRef.scrollIntoView(), 0);
  }, { defer: false }));

  return (
    <>
    <Show when={visible()}>
      <ul class={`${styles['viewport-chat-messages-items']} hlp-text-overflow-parent`}>
        <For each={messages().slice(0, messages().length)}
             fallback={<p style="text-align: center;">empty</p>}>
          {(message, _) => (
            <Dynamic component={DynamicChatMesssagesComponents[message.alignment]} {...messages} />
          )}
        </For>
      </ul>

      <div style={{height: "0px"}} ref={props.ref} ref={viewportChatMessagesRef}></div>
    </Show>
    </>
  );
}
```

**Comment by atk** [comment]
That's what https://solid-primitives.netlify.app/package/refs#mergerefs is for.

**Reply by Grabber** [reply]
Inside the child, why is it working?

```
<div style={{height: "0px"}} ref={props.ref} ref={viewportChatMessagesRef}></div>
```

Is it possible to just have one ref and use it inside the child without having to create another local variable and another ref?

```
<div style={{height: "0px"}} ref={props.ref}></div>
```

I don't want to rely on third-party components, I think the core framework should solve it.

**Reply by atk** [reply]
We wouldn't have a third-party component If the core framework did this. You can use a directive instead of a ref - it does very much the same.

**Reply by Grabber** [reply]
What is a directive in this case? Do you know where in the core code I could check why ref ref is working?

**Reply by atk** [reply]
You can simply go to the playground to check the output and search the docs for custom directives.

**Reply by Grabber** [reply]
The question is still: why use mergeRefs if passing refs multiple times in child is working? Is it a bug?

**Comment by Grabber** [comment]
@atk

I think it's a simpler alternative, without relying on third party libs:

## MyParentComponent.jsx
```
export default function MyParentComponent(props) {
  let parentRef;
  
  return (
    <>
    <MyChildComponent ref={parentRef} />
    </>
  );
}
```

## MyChildComponent.jsx
```
export default function MyChildComponent(props) {
  let childRef;

  createEffect(() => {
    // do something with `localRef`...
  }, { defer: false });

  const mergeRefs = (...refs) => {
    return (r) => { refs.forEach((ref) => ref(r)); }
  }

  return (
    <>
    <div ref={mergeRefs(props.ref, el => childRef = el)}></div>
    </>
  );
}
```

`parentRef` is fowarded from parent to child components as `props.ref` and merged with `childRef` as `ref`
